Improve the error message when using st.memo with Snowpark session

sfc-gh-tszerszen opened the pull request:
Improve the error message when using st.memo with Snowpark session. Tell the dev to use st.singleton instead.

<!--
Before contributing (PLEASE READ!)

⚠️ If your contribution is more than a few lines of code, then prior to starting to code on it please post in the issue saying you want to volunteer, then wait for a positive response. And if there is no issue for it yet, create it first.

This helps make sure:

  1. Two people aren't working on the same thing
  2. This is something Streamlit's maintainers believe should be implemented/fixed
  3. Any API, UI, or deeper architectural changes that need to be implemented have been fully thought through by Streamlit's maintainers
  4. Your time is well spent!

More information in our wiki: https://github.com/streamlit/streamlit/wiki/Contributing
-->

## 📚 Context

_Please describe the project or issue background here_

- What kind of change does this PR introduce?

  - [ ] Bugfix
  - [ ] Feature
  - [ ] Refactoring
  - [x] Other, please describe:

## 🧠 Description of Changes

- _Add bullet points summarizing your changes here_

  - [ ] This is a breaking API change
  - [ ] This is a visible (user-facing) change

**Revised:**

_Insert screenshot of your updated UI/code here_

**Current:**

_Insert screenshot of existing UI/code here_

## 🧪 Testing Done

- [ ] Screenshots included
- [ ] Added/Updated unit tests
- [ ] Added/Updated e2e tests

## 🌐 References

_Does this depend on other work, documents, or tickets?_

- **Issue**: Closes #XXXX

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-kjavadyan commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -248,7 +248,28 @@ def get_or_create_cached_value():
+                        and type(return_value).__name__ == "Session"
+                    ):
+
+                        class SnowparkSessionInMemoDecoratorError(TypeError):
```
I think it would be great to move `SnowparkSessionInMemoDecoratorError` class definition to https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/errors.py

sfc-gh-kjavadyan commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -248,7 +248,28 @@ def get_or_create_cached_value():
+                                TypeError.__init__(
+                                    self,
+                                    """
+                                    The function snowflake.snowpark.session.Session is cached with st.memo,
```
Let's try to keep wording consistent with this doc: https://www.notion.so/streamlit/Spec-9497a30d456f474eaa2343032ef8cb48#2153548903b14c89b0eede7530d0ca06

Final wording you can discuss with @jrieke !

sfc-gh-tszerszen commented:
@sfc-gh-kjavadyan thank you for the review, I've just introduced your suggestions 👍 I'll get to @jrieke for the next review 👍

sfc-gh-kbregula commented on lib/streamlit/runtime/caching/cache_utils.py:
```diff
@@ -248,7 +248,17 @@ def get_or_create_cached_value():
+                    module_name = (
+                        f"{return_value.__module__}.{type(return_value).__name__}"
+                    )
+                    if module_name == "snowflake.snowpark.session.Session":
```
Can you use `type_util.is_type` function?

sfc-gh-tszerszen replied:
@sfc-gh-kbregula thank you for the review 👍 I've used `type_util.is_type` here 👍

sfc-gh-kbregula commented:
Can you add tests to avoid regression?

sfc-gh-tszerszen commented:
@sfc-gh-kjavadyan 🤔 I've reversed changes in `errors.py` and I wonder if it makes sense to create new errors, as I've done before? They will be used only once, and I've 👀 seen it will be good to use generic `errors.MarkdownFormattedException` which will allow error messages to be nicely formatted

sfc-gh-tszerszen commented:
> Can you add tests to avoid regression?

Sure 👍 The question is 🤔 should not we wait for setting up the CI before we write the tests? I guess we need to be able to create a connection to Snowflake to make them, or am I mistaken?

jrieke commented:
> > Can you add tests to avoid regression?
> 
> Sure 👍 The question is 🤔 should not we wait for setting up the CI before we write the tests? I guess we need to be able to create a connection to Snowflake to make them, or am I mistaken?

Drive-by comment from someone not intimately familiar with the code base:I think it's enough if we test this with other unserializable objects. Doesn't need to be a Snowpark connection. But please ask an eng to confirm this, maybe I'm talking BS :D

jrieke commented:
Already wrote @sfc-gh-tszerszen on Slack but just FYI that I updated the spec a bit, see [here](https://www.notion.so/streamlit/Spec-9497a30d456f474eaa2343032ef8cb48#6e4bcb6c93cf4519be604cecfacf0de1).

sfc-gh-kbregula commented:
@sfc-gh-tszerszen If you don't have credentials, we shouldn't run tests. If you have credentials locally, the tests should be able to run. I am working on configuring Github Action. I am working on configuring GitHub Action. I am also working on a document which I will describe exactly as if I wanted to run tests locally and on Github Action.

sfc-gh-tszerszen committed:
Introduce UnserializableReturnValueError in st.experimental_memo

sfc-gh-tszerszen commented:
> > > Can you add tests to avoid regression?
> > 
> > 
> > Sure 👍 The question is 🤔 should not we wait for setting up the CI before we write the tests? I guess we need to be able to create a connection to Snowflake to make them, or am I mistaken?
> 
> Drive-by comment from someone not intimately familiar with the code base:I think it's enough if we test this with other unserializable objects. Doesn't need to be a Snowpark connection. But please ask an eng to confirm this, maybe I'm talking BS :D

You're right 👍 We've come to this from Snowpark integration point of view, however the solution is general and includes every kind of integration error, hence I've introduced general test case, which I think covers it cc: @sfc-gh-kbregula 





> Already wrote @sfc-gh-tszerszen on Slack but just FYI that I updated the spec a bit, see [here](https://www.notion.so/streamlit/Spec-9497a30d456f474eaa2343032ef8cb48#6e4bcb6c93cf4519be604cecfacf0de1).

Thank you for this detailed spec, I've updated this PR accordingly and it's ready I think, please re-review it:
R: @jrieke @sfc-gh-kbregula

sfc-gh-tszerszen committed:
improve tests

sfc-gh-tszerszen committed:
update cached func

jrieke reviewed (approved):
LGTM from product perspective
